### PR TITLE
Issue 52 feat/api create game service

### PR DESCRIPTION
**Pull request by KonTra7cf**

### CODE REVIEW

**Review by Veryyapeee — APPROVED** [review]

File: src/common/middlewares/game.middleware.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import { StatusCodes, ReasonPhrases } from 'http-status-codes';
+import { GameService } from '../../game';
+import { Middleware } from './middleware.type';
+
+export const gameMiddleware: Middleware = async (req, res, next) => {
```
Comment (PoProstuMieciek) [inline src/common/middlewares/game.middleware.ts:5]: Why using middleware? It just doesn't seem right really. @Veryyapeee let me know if this is needed. I think the cleaner solution would be checking it in request.

**Review by PoProstuMieciek — CHANGES_REQUESTED** [review]

File: src/index.ts [hunk 2]
```diff
@@ -3,19 +3,21 @@ import { WinstonAdaptor } from 'typeorm-logger-adaptor/logger/winston';
...
+import { API, funnyHeaderMiddleware, gameMiddleware } from './common';
 import { ReportController } from './report';
 import { AuthController } from './auth';
 import { UserController } from './user';
+import { GameController } from './game';
 import { FeedController } from './feed/feed.controller';
 
 const server = new API({
-    middlewares: [funnyHeaderMiddleware],
+    middlewares: [funnyHeaderMiddleware, gameMiddleware],
```
Comment (PoProstuMieciek) [inline src/index.ts:14]: We don't want to use this middleware globally. It will probably break other endpoints.

**Review by PoProstuMieciek — CHANGES_REQUESTED** [review]

File: src/game/game.controller.ts [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        super('/game');
+
+        const router = this.getRouter();
+
+        router.get('/', this.getAllGames);
+        router.post('/', this.addGame);
+        router.get('/:id', this.getGame);
+        router.delete('/:id', this.removeGame);
+        router.post('/:id/experience', this.addExperienceLevel);
+        router.delete('/:id/experience/:lvl_id', this.removeExperienceLevel);
```
Comment (PoProstuMieciek) [inline src/game/game.controller.ts:19]: maybe do `/:id/level/:lvl_id`

File: src/game/game.controller.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    async getAllGames(req: Request, res: Response) {
+        return res.send(await GameService.getAllGames());
+    }
+
+    async getGame(req: Request, res: Response) {
+        const id = req.params.id;
+
+        try {
```
Comment (PoProstuMieciek) [inline src/game/game.controller.ts:29]: remove the try catch block
no need to catch errors here
i already did this globally

File: src/game/game.controller.ts [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+        const added = await GameService.addGame(body);
+
+        return res.status(StatusCodes.CREATED).send(instanceToPlain(added));
+    }
+
+    async removeGame(req: Request, res: Response) {
+        const id = req.params.id;
+
+        try {
```
Comment (PoProstuMieciek) [inline src/game/game.controller.ts:59]: remove try catch

File: src/game/game.controller.ts [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    async getGame(req: Request, res: Response) {
+        const id = req.params.id;
+
+        try {
+            const game = await GameService.getGame(id);
+
+            if (!game) {
+                return res.status(StatusCodes.NOT_FOUND).send('Game not found');
```
Comment (PoProstuMieciek) [inline src/game/game.controller.ts:33]: also please remove the text from .send()
`Game not found` is not needed. we have a status code that tells us the same

File: src/game/game.controller.ts [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    async removeGame(req: Request, res: Response) {
+        const id = req.params.id;
+
+        try {
+            const removed = await GameService.removeGame(id);
+
+            if (!removed) {
+                return res.status(StatusCodes.NOT_FOUND).send('Game not found');
```
Comment (PoProstuMieciek) [inline src/game/game.controller.ts:63]: remove `Game not found`

File: src/game/game.controller.ts [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+            }
+            return res
+                .status(StatusCodes.NO_CONTENT)
+                .send('Game has been removed');
```
Comment (PoProstuMieciek) [inline src/game/game.controller.ts:67]: NO_CONTENT 204 should be just OK 200
also `Game has been removed` is not essential. we have status code

change that to `return res.send();`

File: src/game/game.controller.ts [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+            return res
+                .status(StatusCodes.INTERNAL_SERVER_ERROR)
+                .send(ReasonPhrases.INTERNAL_SERVER_ERROR);
+        }
+    }
+
+    async addExperienceLevel(req: Request, res: Response) {
+        const game_id = req.params.id;
+
+        try {
```
Comment (PoProstuMieciek) [inline src/game/game.controller.ts:79]: remove try catch

File: src/game/game.controller.ts [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+            const game = await GameService.getGame(game_id);
+
+            if (!game) {
+                return res.status(StatusCodes.NOT_FOUND).send('Game not found');
```
Comment (PoProstuMieciek) [inline src/game/game.controller.ts:83]: ~~Game not found~~

File: src/game/game.controller.ts [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+                .status(StatusCodes.INTERNAL_SERVER_ERROR)
+                .send(ReasonPhrases.INTERNAL_SERVER_ERROR);
+        }
+    }
+
+    async removeExperienceLevel(req: Request, res: Response) {
+        const game_id = req.params.id;
+        const lvl_id = req.params.lvl_id;
+
+        try {
```
Comment (PoProstuMieciek) [inline src/game/game.controller.ts:107]: rmeov try catch

File: src/game/game.controller.ts [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+            const game = await GameService.getGame(game_id);
+
+            if (!game) {
+                return res.status(StatusCodes.NOT_FOUND).send('Game not found');
```
Comment (PoProstuMieciek) [inline src/game/game.controller.ts:111]: ~~Game not found~~

File: src/game/game.controller.ts [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+            }
+
+            const removed = await GameService.removeExperienceLevel(lvl_id);
+
+            if (!removed) {
+                return res
+                    .status(StatusCodes.NOT_FOUND)
+                    .send('Experience level not found');
```
Comment (PoProstuMieciek) [inline src/game/game.controller.ts:119]: ~~Experience level not found~~

File: src/game/game.controller.ts [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+            }
+            return res
+                .status(StatusCodes.NO_CONTENT)
+                .send('Experience level has been removed');
```
Comment (PoProstuMieciek) [inline src/game/game.controller.ts:123]: ~~NO_CONTENT~~ **OK**
~~Experience level has been removed~~
**`return res.send();`**

File: src/game/game.service.ts [hunk 15]
```diff
@@ -0,0 +1,58 @@
+import { Game, ExperienceLevel, AddGameDto, AddLevelDto } from '.';
+
+export const GameService = new (class {
+    async getAllGames() {
+        const games = await Game.find();
+
+        if (!games.length) {
```
Comment (PoProstuMieciek) [inline src/game/game.service.ts:7]: `if` not needed
we will always get an array or an empty array

File: src/game/game.controller.ts [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+    constructor() {
+        super('/game');
+
+        const router = this.getRouter();
+
+        router.get('/', this.getAllGames);
+        router.post('/', this.addGame);
+        router.get('/:id', this.getGame);
+        router.delete('/:id', this.removeGame);
+        router.post('/:id/experience', this.addExperienceLevel);
```
Comment (PoProstuMieciek) [inline src/game/game.controller.ts:18]: `router.get('/:id/experience', this.getExperienceLevels);`
one endpoint is missing

**Review by PoProstuMieciek — CHANGES_REQUESTED** [review]
everything is almost perfect

**quick summary:**
one endpoint (for listing game levels) is missing
try catch blocks are not required
we don't need to send additional text to report errors. http codes are enough
HTTP NO_CONTENT is not perfect in this case. just use HTTP OK

File: src/game/game.controller.ts [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+        const game_id = req.params.id;
+
+        const game = await GameService.getGame(game_id);
+
+        if (!game) {
+            return res
+                .status(StatusCodes.NOT_FOUND)
+                .send(ReasonPhrases.NOT_FOUND);
+        }
+        return res.send(await GameService.getExperienceLevels());
```
Comment (PoProstuMieciek) [inline src/game/game.controller.ts:75]: you need to get levels of the specified game

now you are printing every level despite being given a game
you need to somewhat pass the game into GameService.getExperienceLevels()

**Review by PoProstuMieciek — CHANGES_REQUESTED** [review]

File: src/game/game.service.ts [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+        }
+
+        return game.remove();
+    }
+
+    async getExperienceLevels(gameId: string) {
+        const levels = await Game.find({
+            relations: ['levels'],
+            where: {
+                id: `${gameId}`
```
Comment (PoProstuMieciek) [inline src/game/game.service.ts:42]: id: gameId

**Review by PoProstuMieciek — APPROVED** [review]
if it works, then lets merge

File: src/game/game.controller.ts [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+    async getExperienceLevels(req: Request, res: Response) {
+        const game_id = req.params.id;
+
+        const game = await GameService.getGame(game_id);
+
+        if (!game) {
+            return res
+                .status(StatusCodes.NOT_FOUND)
+                .send(ReasonPhrases.NOT_FOUND);
+        }
```
Comment (PoProstuMieciek) [inline src/game/game.controller.ts:74]: remove useless lines 68-74:
```ts
const game = await GameService.getGame(game_id);

        if (!game) {
            return res
                .status(StatusCodes.NOT_FOUND)
                .send(ReasonPhrases.NOT_FOUND);
        }
```

**Review by PoProstuMieciek — CHANGES_REQUESTED** [review]

**Review by PoProstuMieciek — APPROVED** [review]

**Pull request merged**